feat(keyboard): add chinese input function to keyboard

huenrong opened the pull request:
### Description of the feature or fix

Add a button to the keyboard to switch the input method (Chinese and English) #5598

FASTSHIFT commented:
Please fix compile warnings based on [CI information](https://github.com/lvgl/lvgl/pull/5779/checks).
![image](https://github.com/lvgl/lvgl/assets/26767803/77d5c354-a841-419e-b991-685c3c6f189d)

huenrong commented:
@FASTSHIFT Hi, I solved some problems, but some checks are still unsuccessful. Is it because I modified the layout of the keyboard?

old keyboard layout
![old keyboard](https://github.com/lvgl/lvgl/assets/47348530/b3e46c0b-6585-42aa-821b-ba7ca00181c8)
new keyboard layout
![new keyboard](https://github.com/lvgl/lvgl/assets/47348530/594339b8-55c0-4e3e-a3c8-dcf5d11610db)

Modify content:
- Added a button for switch the input method (Chinese and English)
- Reduced keyboard height and added candidate area

![keyboard_zh](https://github.com/lvgl/lvgl/assets/47348530/efd3077c-865c-4d71-96c1-58449f161fb8)

FASTSHIFT commented:
You can try deleting the old reference image and re-running the [test](https://github.com/lvgl/lvgl/blob/master/tests/README.md) to generate a new reference image.

FASTSHIFT commented on src/widgets/keyboard/lv_keyboard.c:
```diff
@@ -243,6 +400,23 @@ void lv_keyboard_set_map(lv_obj_t * obj, lv_keyboard_mode_t mode, const char * m
 }
 
+#if LV_USE_KEYBOARD_PINYIN
+void lv_keyboard_set_font(lv_obj_t * obj, const lv_font_t * font)
```
This may not be needed

huenrong replied:
> This may not be needed

This function(`lv_keyboard_set_font()`) is to set the font of candidate Chinese characters. Some Chinese characters cannot be displayed in the default font (lv_font_simsun_16_cjk), Just like the picture I sent earlier, some Chinese characters are displayed as square boxes.

FASTSHIFT replied:
I mean it can be set outside:
```c
lv_obj_t * kb = lv_keyboard_create(lv_scr_act());
lv_obj_set_style_text_font(kb, my_font, 0);
```
and there is no need to wrap this function with a layer.

huenrong replied:
> I mean it can be set outside:
> 
> ```c
> lv_obj_t * kb = lv_keyboard_create(lv_scr_act());
> lv_obj_set_style_text_font(kb, my_font, 0);
> ```
> 
> and there is no need to wrap this function with a layer.

You are right, i will delete this function

huenrong replied:
I have deleted this function(`lv_keyboard_set_font()`)

XuNeo commented on Kconfig:
```diff
@@ -928,6 +928,18 @@ menu "LVGL configuration"
 		config LV_USE_KEYBOARD
 			bool "Keyboard."
 			default y if !LV_CONF_MINIMAL
+		config LV_USE_KEYBOARD_PINYIN
```
I though it's a new widget but from the code, it's an function extension to existing keyboard widget. How about calling is `LV_KEYBOARD_PINYIN`?

huenrong replied:
> I though it's a new widget but from the code, it's an function extension to existing keyboard widget. How about calling is `LV_KEYBOARD_PINYIN`?

ok

XuNeo commented on src/widgets/keyboard/lv_keyboard_pinyin.c:
```diff
@@ -0,0 +1,999 @@
+
+/**
+ * @file lv_keyboard_pinyin.c
```
Would `lv_pinyin_default_dict.c/.h` be better?

huenrong replied:
> Would `lv_pinyin_default_dict.c/.h` be better?

Whether the file name needs to be related to the folder name, such as prefixed with `lv_keyboard_`, But `lv_keyboard_pinyin_default_dict.c/.h` is too long

XuNeo commented on src/widgets/keyboard/lv_keyboard.c:
```diff
@@ -388,14 +899,147 @@ static void lv_keyboard_constructor(const lv_obj_class_t * class_p, lv_obj_t * o
+    lv_color_t keyboard_bg_color = lv_obj_get_style_bg_color(obj, 0);
+
+    static lv_style_t pinyin_default_style;
+    lv_style_init(&pinyin_default_style);
```
style should only be initialized once if I recall correctly.

huenrong replied:
> style should only be initialized once if I recall correctly.

Is this okay?
```c
lv_color_t keyboard_bg_color = lv_obj_get_style_bg_color(obj, 0);

static lv_style_t pinyin_default_style;
if(pinyin_default_style.prop_cnt < 1) {
    lv_style_init(&pinyin_default_style);
    lv_style_set_bg_opa(&pinyin_default_style, keyboard_bg_opa);
    lv_style_set_bg_color(&pinyin_default_style, keyboard_bg_color);
    lv_style_set_border_width(&pinyin_default_style, 0);
    lv_style_set_pad_all(&pinyin_default_style, 0);
    lv_style_set_pad_gap(&pinyin_default_style, 0);
    lv_style_set_radius(&pinyin_default_style, 0);
#if LV_FONT_SIMSUN_16_CJK
    lv_style_set_text_font(&pinyin_default_style, &lv_font_simsun_16_cjk);
#endif
}
```

FASTSHIFT replied:
Can we make this style the [default theme](https://github.com/lvgl/lvgl/blob/master/src/themes/default/lv_theme_default.c#L1041)?

huenrong replied:
> Can we move this style to the [default theme](https://github.com/lvgl/lvgl/blob/master/src/themes/default/lv_theme_default.c#L1041)?

Can anyone please confirm this problem? If not, I will modify it according to the previous plan.

XuNeo commented on src/widgets/keyboard/lv_keyboard.c:
```diff
@@ -371,14 +582,314 @@ void lv_keyboard_def_event_cb(lv_event_t * e)
+    if(*pinyin_str == ' ')     return NULL;
+
+    size_t pinyin_str_len = strlen(pinyin_str);
+    char new_pinyin_str[pinyin_str_len + 1];
```
Somebody reported this line may not  compile on MSVC. Maybe we can use malloc here? Is `pinyin_str` max length known?

huenrong replied:
> Somebody reported this line may not compile on MSVC. Maybe we can use malloc here? Is `pinyin_str` max length known?

The code does not limit the max length of `pinyin_str`, `pinyin_str` comes from label(`lv_label_get_text()`).

XuNeo commented on src/widgets/keyboard/lv_keyboard.c:
```diff
@@ -371,14 +582,314 @@ void lv_keyboard_def_event_cb(lv_event_t * e)
+        }
+    }
+
+    if(keyboard->dict == NULL) {
```
Move to line 652 to exit earlier.

huenrong replied:
> Move to line 652 to exit earlier.

okay

XuNeo commented on src/widgets/keyboard/lv_keyboard.c:
```diff
@@ -371,14 +582,314 @@ void lv_keyboard_def_event_cb(lv_event_t * e)
+        return NULL;
+    }
+
+    lv_keyboard_pinyin_dict_t * pinyin_head = (lv_keyboard_pinyin_dict_t *)keyboard->dict[*new_pinyin_str - 'a'];
```
It looks like no need to cast here.
```suggestion
    lv_keyboard_pinyin_dict_t * pinyin_head = keyboard->dict[*new_pinyin_str - 'a'];
```

huenrong replied:
> It looks like no need to cast here.

`keyboard->dict[*new_pinyin_str - 'a']` is `const lv_keyboard_pinyin_dict_t *`, `pinyin_head` is `lv_keyboard_pinyin_dict_t *` I got a warning when compiling on Ubuntu 20.04

XuNeo replied:
How about `const lv_keyboard_pinyin_dict_t * pinyin_head = keyboard->dict[*new_pinyin_str - 'a'];`

huenrong replied:
> How about `const lv_keyboard_pinyin_dict_t * pinyin_head = keyboard->dict[*new_pinyin_str - 'a'];`

okay

XuNeo commented on src/widgets/keyboard/lv_keyboard.c:
```diff
@@ -371,14 +582,314 @@ void lv_keyboard_def_event_cb(lv_event_t * e)
+    size_t i = 0;
+    for(; pinyin_head < pinyin_tail; pinyin_head++) {
+        for(i = 0; i < pinyin_str_len - 1; i++) {
+            if((*pinyin_head).pinyin_list == NULL) return NULL;
```
```suggestion
            if(pinyin_head->pinyin_list == NULL) return NULL;
```

huenrong replied:
> 

Thanks, for ease of operation, I will adjust it locally

XuNeo commented on src/widgets/keyboard/lv_keyboard.c:
```diff
@@ -371,14 +582,314 @@ void lv_keyboard_def_event_cb(lv_event_t * e)
+        for(i = 0; i < pinyin_str_len - 1; i++) {
+            if((*pinyin_head).pinyin_list == NULL) return NULL;
+
+            if(*((*pinyin_head).pinyin_list + i) != new_pinyin_str[i + 1]) break;
```
Is it equivalent to below?
```suggestion
            if(pinyin_head->pinyin_list[i] != new_pinyin_str[i + 1]) break;
```

huenrong replied:
> Is it equivalent to below?

Thanks, for ease of operation, I will adjust it locally

XuNeo commented on src/widgets/keyboard/lv_keyboard.c:
```diff
@@ -371,14 +582,314 @@ void lv_keyboard_def_event_cb(lv_event_t * e)
+        /*perfect match*/
+        if(i == (pinyin_str_len - 1)) {
+            /*The Chinese character in UTF-8 encoding format is 3 bytes*/
+            *candidate_num = strlen((const char *)(*pinyin_head).pinyin_mb_list) / 3;
```
ditto.

huenrong replied:
> ditto.

okay

XuNeo commented on src/widgets/keyboard/lv_keyboard.c:
```diff
@@ -371,14 +582,314 @@ void lv_keyboard_def_event_cb(lv_event_t * e)
+static void lv_keyboard_proc_pinyin(lv_keyboard_t * keyboard)
+{
+    char * input_pinyin = lv_label_get_text(keyboard->pinyin_label);
+    if(lv_strlen(input_pinyin) == 0) {
```
if (input_pinyin == NULL || lv_strlen(input_pinyin) == 0)

huenrong replied:
> if (input_pinyin == NULL || lv_strlen(input_pinyin) == 0)

okay

XuNeo commented on src/widgets/keyboard/lv_keyboard.c:
```diff
@@ -371,14 +582,314 @@ void lv_keyboard_def_event_cb(lv_event_t * e)
+    }
+
+    size_t candidate_num = 0;
+    char * candidate_str = lv_keyboard_search_chinese(&candidate_num, keyboard, input_pinyin);
```
Check NULL before using it below.

huenrong replied:
> Check NULL before using it below.

okay
```c
size_t candidate_num = 0;
char * candidate_str = lv_keyboard_search_chinese(&candidate_num, keyboard, input_pinyin);
if(candidate_str == NULL || lv_strlen(candidate_str) == 0 || candidate_num == 0) return; // add
```

XuNeo commented on src/widgets/keyboard/lv_keyboard.c:
```diff
@@ -371,14 +582,314 @@ void lv_keyboard_def_event_cb(lv_event_t * e)
+    }
+
+    for(size_t i = 0; i < candidate_num; i++) {
+        keyboard->candidate_btnm_map[i] = (char *)lv_malloc(4 * sizeof(char));
```
Let make the code shorter by define a var here.

```c
char * str = lv_malloc(4);
LV_ASSERT_NULL(str);
lv_memcpy(str, candidate_str + i * 3, 3);
str[3] = '\0';
keyboard->candidate_btnm_map[i] = str;
```

huenrong replied:
> Let make the code shorter by define a var here.
> 
> ```c
> char * str = lv_malloc(4);
> LV_ASSERT_NULL(str);
> lv_memcpy(str, candidate_str + i * 3, 3);
> str[3] = '\0';
> keyboard->candidate_btnm_map[i] = str;
> ```

okay

XuNeo commented on src/widgets/keyboard/lv_keyboard.c:
```diff
@@ -371,14 +582,314 @@ void lv_keyboard_def_event_cb(lv_event_t * e)
+
+    lv_obj_remove_flag(keyboard->ext_btn, LV_OBJ_FLAG_HIDDEN);
+
+    int32_t width = (btnm_width / keyboard->candidate_num);
```
Please confirm if it's needed to check divide zero here. And similar for below division.

huenrong replied:
> Please confirm if it's needed to check divide zero here. And similar for below division.

No need to check here, But add the following check
```c
size_t candidate_num = 0;
char * candidate_str = lv_keyboard_search_chinese(&candidate_num, keyboard, input_pinyin);
if(candidate_str == NULL || lv_strlen(candidate_str) == 0 || candidate_num == 0) return; // add
```

huenrong commented:
@FASTSHIFT @XuNeo 
Hi, I have solved the above problem and re-push.
A bad news, there is a bug in Use extend candidate button matrix (`#define LV_KEYBOARD_PINYIN_USE_EXT_BUTTONMATRIX 1`), should I continue to change it in this PR?

100ask commented:
> ### Description of the feature or fix
> Add a button to the keyboard to switch the input method (Chinese and English) #5598

I don't think it's a good idea to switch between Chinese and English in lv_keyboard. If there are more input methods from other languages, it will lead to lv_keyboard being bloated.
Why not implement it on the basis of lv_ime_pinyin(https://docs.lvgl.io/master/others/ime_pinyin.html) (more input methods such as lv_ime_xxx)?

huenrong commented:
> > ### Description of the feature or fix
> > Add a button to the keyboard to switch the input method (Chinese and English) #5598
> 
> hi,
> 
> I don't think it's a good idea to switch between Chinese and English in lv_keyboard. If there are more input methods from other languages, it will lead to lv_keyboard being bloated.
> 
> Why not implement it on the basis of [lv_ime_pinyin](https://docs.lvgl.io/master/others/ime_pinyin.html) (more input methods such as lv_ime_xxx)?
> 
> reference:
> 
> * https://keyshorts.com/blogs/blog/44712961-how-to-identify-laptop-keyboard-localization
> * https://en.wikipedia.org/wiki/List_of_QWERTY_keyboard_language_variants

Hi, I think it is inconvenient to switch between inputting Chinese and English when using `lv_ime_pinyin`. Maybe I don’t know it well enough.

PGNetHun reviewed (changes requested):
There are too many changes in `lv_keyboard.c`, too many `#if LV_KEYBOARD_PINYIN` , that it is worth to refactor `lv_keyboard.c` to simply register keyboard layouts and functionalities (like registering FS drivers), and separate the default English layout and Pinyin layout.
By separating the keyboard layouts to separate files, we can add later Cyrillic or arabish keyboard layouts easier. (I don't know whether these are already supported).

So, after refactor we would have:
- `lv_keyboard`
- `lv_keyboard_layout_en`
- `lv_keyboard_layout_pinyin`
- `[lv_keyboard_layout_cyrillic]`

Changing between the layout is done by a button.

The aim is clean code (avoid too many `#if` lines), good structure, and extensibility (new files).

Of course some keyboard functionalities can be shared, those can be stored either in `lv_keyboard`, or `lv_keyboard_shared`.

lhdjply commented:
I think we should make changes on [lv_ime_pinyin](https://github.com/lvgl/lvgl/blob/master/src/others/ime/lv_ime_pinyin.c) instead of adding a new one.

kisvegabor commented on src/widgets/keyboard/lv_keyboard.c:
```diff
@@ -49,6 +63,104 @@ const lv_obj_class_t lv_keyboard_class = {
     .name = "keyboard",
 };
 
+#if LV_KEYBOARD_PINYIN
```
It would be more elegant to have something like:
```c
#if LV_USE_ARABIC_PERSIAN_CHARS 
  #define PEERSIAN_KEY  "أب",
  #define PEERSIAN_CTRL  LV_KEYBOARD_CTRL_BUTTON_FLAGS | 2,
#else
  #define PEERSIAN_KEY  
  #define PEERSIAN_CTRL
#endif
//Same for Chines
//..

static const char * const default_en_kb_map_lc[] = {" ", "\n", /*candidate content*/
                                                    "1#", "q", "w", "e", "r", "t", "y", "u", "i", "o", "p", LV_SYMBOL_BACKSPACE, "\n",
                                                    "ABC", "a", "s", "d", "f", "g", "h", "j", "k", "l", LV_SYMBOL_NEW_LINE, "\n",
                                                    "_", "-", "z", "x", "c", "v", "b", "n", "m", ".", ",", ":", "\n",
                                                    LV_SYMBOL_KEYBOARD,  CHINESE_KEY PEERSIAN_KEY  LV_SYMBOL_LEFT, " ", LV_SYMBOL_RIGHT, LV_SYMBOL_OK, ""
                                                   };

```

huenrong replied:
> It would be more elegant to have something like:
> 
> ```c
> #if LV_USE_ARABIC_PERSIAN_CHARS 
>   #define PEERSIAN_KEY  "أب",
>   #define PEERSIAN_CTRL  LV_KEYBOARD_CTRL_BUTTON_FLAGS | 2,
> #else
>   #define PEERSIAN_KEY  
>   #define PEERSIAN_CTRL
> #endif
> //Same for Chines
> //..
> 
> static const char * const default_en_kb_map_lc[] = {" ", "\n", /*candidate content*/
>                                                     "1#", "q", "w", "e", "r", "t", "y", "u", "i", "o", "p", LV_SYMBOL_BACKSPACE, "\n",
>                                                     "ABC", "a", "s", "d", "f", "g", "h", "j", "k", "l", LV_SYMBOL_NEW_LINE, "\n",
>                                                     "_", "-", "z", "x", "c", "v", "b", "n", "m", ".", ",", ":", "\n",
>                                                     LV_SYMBOL_KEYBOARD,  CHINESE_KEY PEERSIAN_KEY  LV_SYMBOL_LEFT, " ", LV_SYMBOL_RIGHT, LV_SYMBOL_OK, ""
>                                                    };
> ```

okay, i will try it

kisvegabor commented on src/widgets/keyboard/lv_keyboard.c:
```diff
@@ -287,58 +454,78 @@ void lv_keyboard_def_event_cb(lv_event_t * e)
         lv_buttonmatrix_set_map(obj, kb_map[LV_KEYBOARD_MODE_TEXT_LOWER]);
         lv_keyboard_update_ctrl_map(obj);
-        return;
+        goto update_candidate;
```
Instead of `goto` let's have a function for this large `if-else if` from which we can clearly return

huenrong replied:
> Instead of `goto` let's have a function for this large `if-else if` from which we can clearly return

okay, i will try it

huenrong commented on docs/widgets/keyboard.rst:
```diff
@@ -35,6 +35,7 @@ The Keyboards have the following modes:
 - :cpp:enumerator:`LV_KEYBOARD_MODE_TEXT_UPPER` Display upper case letters
 - :cpp:enumerator:`LV_KEYBOARD_MODE_TEXT_SPECIAL` Display special characters
 - :cpp:enumerator:`LV_KEYBOARD_MODE_NUMBER` Display numbers, +/- sign, and decimal dot
+- :cpp:enumerator:`LV_KEYBOARD_MODE_PINYIN` Add Zh and En button to toggles input chinese and english
```
> Grammar fix: _to toggles_ -> _to toggle_

okay

huenrong commented on lv_conf_template.h:
```diff
@@ -516,6 +516,18 @@
 
+/*1: Enable Pinyin input method*/
+/*Requires: lv_keyboard*/
+#define LV_KEYBOARD_PINYIN 1
```
> I think we should **not** enable Pinyin by default. For most people english keyboard is enough, only chinese developers need pinyin keyboard, too.

okay

huenrong commented on src/lv_conf_internal.h:
```diff
@@ -1622,6 +1622,48 @@
+            #define LV_KEYBOARD_PINYIN 0
+        #endif
+    #else
+        #define LV_KEYBOARD_PINYIN 1
```
> I think we should **not** enable Pinyin by default.

okay

huenrong commented on tests/ref_imgs/widgets/keyboard_1.png:
> Why is the height changed in ref images? It should not change, because on smaller displays it could be very small. Please do not change the original visual layout.

Maybe the ref images was changed because I mistakenly enabled LV KEYBOARD PINYIN by default, I'll check it

huenrong commented:
@kisvegabor @PGNetHun 
Hi, I refactor the pinyin function, I think it is clearer than the last commit, please review.
If we need to separate Arabic (Persian) and Pinyin from the keyboard, I think it is better to start a new PR

kisvegabor commented:
@PGNetHun 
> By separating the keyboard layouts to separate files, we can add later Cyrillic or arabish keyboard layouts easier. (I don't know whether these are already supported).

It makes a lot of sense. Besides that I think there is a lot of room for improvements in the keyboard: popovers to select relates characters (like on smart phones), emoji support, better layout/language selector, T9-like suggestions.

We will have a release on Tuesday (March 19), so if merge this PR, and we start updating the keyboard later we will have a lot of braking changes. In light of that I suggest not merging this PR now but open a [future planning issue](https://github.com/lvgl/lvgl/issues/new?assignees=&labels=&projects=&template=feat-planning.yml) to discuss how we can improve the keyboard in general.

kisvegabor commented:
Keeping it open until a feature planning issue is opened.
